### PR TITLE
[Feature] Support CUB dataset

### PR DESCRIPTION
**Pull request by okotaku**
## Motivation

Support [CUB dataset](http://www.vision.caltech.edu/visipedia/CUB-200-2011.html).

Augmentation and optimizer is based on [A Novel Plug-in Module for Fine-Grained Visual Classification](https://paperswithcode.com/paper/a-novel-plug-in-module-for-fine-grained-1).

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by Ezra-Yu** [comment]
Thank you for your awesome contribution, I will run some experiments to get several checkpoints and logs. That would take a little bit of time.

File: mmcls/datasets/cub.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        self.image_class_labels_file = image_class_labels_file
+        self.train_test_split_file = train_test_split_file
+        super(CUB, self).__init__(*args, ann_file=ann_file, **kwargs)
+
+    def load_annotations(self):
+        with open(self.ann_file) as f:
+            samples = [x.strip().split(' ')[1] for x in f.readlines()][:100]
+
+        with open(self.image_class_labels_file) as f:
+            gt_labels = [int(x.strip().split(' ')[1])
```
Comment (Ezra-Yu) [inline mmcls/datasets/cub.py:101]:
```suggestion
            gt_labels = [int(x.strip().split(' ')[1]) - 1
```

the original labels are from 1 to 200, training the CUB full dataset will get an error.: https://stackoverflow.com/questions/51691563/cuda-runtime-error-59-device-side-assert-triggered.

File: mmcls/datasets/cub.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                         for x in f.readlines()][:100]
```
Comment (Ezra-Yu) [inline mmcls/datasets/cub.py:102]: please remove all the `[:100]` in the code.

File: mmcls/datasets/cub.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import numpy as np
+
+from .base_dataset import BaseDataset
+from .builder import DATASETS
+
+
+@DATASETS.register_module()
+class CUB(BaseDataset):
+    """`CUB-200-2011.
```
Comment (Ezra-Yu) [inline mmcls/datasets/cub.py:10]: please modify this line refer to https://github.com/open-mmlab/mmclassification/blob/a7f8e96b31c10ab3e9c133293ca406e6e548475b/mmcls/datasets/cifar.py#L17

File: mmcls/datasets/cub.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        with open(self.image_class_labels_file) as f:
+            gt_labels = [
+                int(x.strip().split(' ')[1]) - 1 for x in f.readlines()
+            ]
+
+        with open(self.train_test_split_file) as f:
+            splits = [int(x.strip().split(' ')[1]) for x in f.readlines()]
+
+        assert len(samples) == len(gt_labels) == len(splits),\
+            'samples, gt_labels and splits should have same length.'
```
Comment (Ezra-Yu) [inline mmcls/datasets/cub.py:109]: Using `f'samples({len(samples)}), gt_labels({len(gt_labels)}) and splits({len(splits)}) should have the same length.'` could show the more detailed messages to users.

**Comment by Ezra-Yu** [comment]
I have reproduced this baseline `r50`(88.2) and `swin-l`(91.9) result in [A Novel Plug-in Module for Fine-Grained Visual Classification](https://paperswithcode.com/paper/a-novel-plug-in-module-for-fine-grained-1).

Could I commit the configs and doc to your branch directly?

**Comment by okotaku** [comment]
@Ezra-Yu Sure.

File: mmcls/datasets/cub.py (hunk 3 above)
Comment (okotaku) [reply mmcls/datasets/cub.py:10]: I'm getting this problem with docformatter's auto-formatting. What should I do?

**Comment by Ezra-Yu** [comment]
I can not commit to your branch directly, maybe you should check the `Allow edits and access to secrets by maintainers`.

for example:
image

**Comment by okotaku** [comment]
@Ezra-Yu I could not run `Allow edits and access to secrets by maintainers` because it was not forked with my personal account. Instead, I gave you write permission and invited as collaborators. Please make sure this is okay.

**Review by Ezra-Yu — APPROVED** [review]
LGTM.

**Comment by okotaku** [comment]
@Ezra-Yu 

I noticed that the following url for the weight is NOT FOUND.
https://download.openmmlab.com/mmclassification/v0/resnet/resnet50_mill_3rdparty_in21k_20220307-bdb3a68b.pth

**Comment by okotaku** [comment]
@Ezra-Yu 

I noticed that the following url for the weight is NOT FOUND.
https://download.openmmlab.com/mmclassification/v0/resnet/resnet50_mill_3rdparty_in21k_20220307-bdb3a68b.pth

**Comment by Ezra-Yu** [comment]
@okotaku  Yes, the weights and logs have not been uploaded to the server, I will urge this to be done.

**Comment by okotaku** [comment]
@Ezra-Yu @mzr1996 This PR was released in v0.22.0, but weights and logs do not yet exist.

**Comment by mzr1996** [comment]
> @Ezra-Yu @mzr1996 This PR was released in v0.22.0, but weights and logs do not yet exist.

My bad, I will upload it soon.

**Comment by mzr1996** [comment]
> @Ezra-Yu @mzr1996 This PR was released in v0.22.0, but weights and logs do not yet exist.

Done.
